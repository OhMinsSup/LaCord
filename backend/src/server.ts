import * as Koa from 'koa';
import * as koaBody from 'koa-body';
import * as compress from 'koa-compress';
import routes from './router';
import database from'./database/db';
import tokenMiddleware from './lib/middlewares/tokenMiddleware';
import cors from './lib/middlewares/cors';

class Server {
    public app: Koa;

    constructor() {
        this.app = new Koa();
        this.initializeDb();
        this.middleware();
        this.routes();
    }
    
    private initializeDb(): void {
        if (!database.connected) {
            database.connect();
            console.log('LaCord Database Conntection âœ…')
        } 
    }

    private middleware(): void {
        const { app } = this;

        app.use(cors);
        app.use(tokenMiddleware);
        app.use(koaBody({
            multipart: true
        }));
        app.use(compress({
            filter: (contentType) => {
                return /text/i.test(contentType)
            },
            threshold: 2048,
            flush: require('zlib').Z_SYNC_FLUSH
        }));
    }

    private routes(): void {
        const { app } = this;
        app.use(routes.routes()).use(routes.allowedMethods());
    }
}

export default new Server().app;
