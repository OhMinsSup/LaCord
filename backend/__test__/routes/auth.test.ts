import server from '../../src/server';
import * as request from 'supertest';
import "reflect-metadata";

describe('Auth Routes', () => {
    jest.useFakeTimers();
    describe('POST: /auth/register/local', () => {
        it('200 Register', async (done) => {
            try {
                const res = await request(server.callback())
                .post('/auth/register/local')
                .send({
                    username: 'test',
                    email: 'test@naver.com',
                    password: '123123'
                });
    
                expect(res.status).toEqual(200);
                expect(Object.keys(res.body.user)).toEqual(
                    expect.arrayContaining(['id','username','email','thumbnail'])
                );
                expect(res.body).toMatchSnapshot();
            } catch (e) {
                done(e)
            }
        });

        it('404 Validate Error', async (done) => {
            try {
                const res = await request(server.callback())
                .post('/auth/register/local')
                .send({
                    username: 123123,
                    email: 'test@naver.com',
                    password: '123123'
                });

                expect(res.status).toEqual(404);
                expect(res.body).toMatchSnapshot();
            } catch (e) {
                done(e);
            }
        });

        it('409 Duplicated_Account', async (done) => {
            try {
                const res = await request(server.callback())
                .post('/auth/register/local')
                .send({
                    username: 'test',
                    email: 'test@naver.com',
                    password: '123123'
                });

                expect(res.status).toEqual(409);
                expect(res.body).toEqual(
                    expect.objectContaining({
                        name: expect.stringMatching('DUPLICATED_ACCOUNT'),
                        payload: expect.any(String)
                    })
                );
                expect(res.body).toMatchSnapshot();
            } catch (e) {
                done(e);
            }
        })
    })
});