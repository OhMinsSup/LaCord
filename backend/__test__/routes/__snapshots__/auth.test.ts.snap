// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth Routes POST: /auth/register/local 404 Validate Error 1`] = `
Object {
  "_object": Object {
    "email": "test@naver.com",
    "password": "123123",
    "username": 123123,
  },
  "details": Array [
    Object {
      "context": Object {
        "key": "username",
        "label": "username",
        "value": 123123,
      },
      "message": "\\"username\\" must be a string",
      "path": Array [
        "username",
      ],
      "type": "string.base",
    },
  ],
  "isJoi": true,
  "name": "ValidationError",
}
`;
